packages/client/auth-core/src/utils/tokenAuth/publicKey.ts

```diff
@@ -0,0 +1,40 @@
+import { decode } from "jsonwebtoken";
+import { JwksClient } from "jwks-rsa";
+
+import PublicKeyCacheProvider from "../PublicKeyCacheProvider";
+
+export async function getFetchedPublicKey(token: string, jwksUri: string) {
+    const decoded = decode(token, { complete: true });
+    if (decoded?.header == null) {
+        throw new Error("Invalid Token: Header Formatted Incorrectly");
+    }
+
+    const signingKey = await getSigningKey(jwksUri, decoded.header.kid);
+    const publicKey = signingKey.getPublicKey();
+    PublicKeyCacheProvider.setCachedPublicKey(publicKey);
+
+    return publicKey;
+}
+
+export async function getCachedPublicKey(token: string, jwksUri: string) {
+    const publicKey = PublicKeyCacheProvider.getCachedPublicKey();
+    if (publicKey != null) {
+        return publicKey;
+    }
+
+    return getFetchedPublicKey(token, jwksUri);
+}
+
+async function getSigningKey(jwksUri: string, kid?: string) {
+    const client = new JwksClient({
+        jwksUri,
+        cache: true,
```

Comment:
If jwks-rsa is using cache, why do we need our own in-memory cache?

Reply:
Now that you mentioned, reading the JWKsClient doc I found this

```
Caching
In order to prevent a call to be made each time a signing key needs to be retrieved you can also configure a cache as follows. If a signing key matching the kid is found, this will be cached and the next time this kid is requested the signing key will be served from the cache instead of calling back to the JWKS endpoint.
```

And now I understand the kid purpose. So agree we shouldn't handle the cache, but we should make the kid dynamic.

I assume the rotation of keys will come with the refresh proposal, so I'll remove the caching strategy from here